zjit/src/hir.rs

```diff
@@ -3561,23 +3561,248 @@ impl Function {
+                );
+            }
+
+            // Get body instructions.
+            for insn_id in &block.insns {
+                let insn_id = self.union_find.borrow().find_const(*insn_id);
+                let insn = self.find(insn_id);
+
+                // Snapshots are not serialized, so skip them.
+                if matches!(insn, Insn::Snapshot {..}) {
+                    continue;
+                }
+
+                // Instructions with no output or an empty type should have an empty type field.
+                let type_str = if insn.has_output() {
+                    let insn_type = self.type_of(insn_id);
+                    if insn_type.is_subtype(types::Empty) {
+                        String::new()
+                    } else {
+                        insn_type.print(&ptr_map).to_string()
+                    }
+                } else {
+                    String::new()
+                };
+
+                let opcode = insn.print(&ptr_map).to_string();
+
+                // Traverse the worklist to get inputs for a given instruction.
+                let mut inputs = VecDeque::new();
+                self.worklist_traverse_single_insn(&self.find(insn_id), &mut inputs);
+                let inputs: Vec<Json> = inputs.into_iter().map(|x| x.0.into()).collect();
+
+                instructions.push(
+                    Self::make_iongraph_instr(
+                        ptr_map.map_id(insn_id.0 as u64) as u64,
+                        insn_id.0,
+                        inputs,
+                        &opcode,
+                        &type_str
+                    )
+                );
+            }
+
+            // Create the block with instructions.
+            let block_ptr = ptr_map.map_id(block_id.0 as u64);
+
+            let (block_successors, block_predecessors) = self.compute_successors_and_predecessors();
+            let predecessors = block_predecessors.get(&block_id)
+                .cloned()
+                .unwrap_or_default();
+            let successors = block_successors.get(&block_id)
+                .cloned()
+                .unwrap_or_default();
+
+            mir_blocks.push(Self::make_iongraph_block(
+                block_ptr as u64,
+                block_id.0 as u64,
+                predecessors,
+                successors,
+                instructions
+            ));
+        }
+
+        Self::make_iongraph_function(pass_name, mir_blocks)
+    }
+
     /// Run all the optimization passes we have.
     pub fn optimize(&mut self) {
+        let mut passes: Vec<Json> = Vec::new();
+
+        if get_option!(dump_hir_iongraph) {
+            passes.push(self.to_iongraph_pass("Unoptimized"));
+        }
+
         // Function is assumed to have types inferred already
         self.type_specialize();
         #[cfg(debug_assertions)] self.assert_validates();
+        if get_option!(dump_hir_iongraph) {
+            passes.push(self.to_iongraph_pass("Type specialize"));
+        }
```

Comment:
The code for each pass seems a bit too verbose now. How about making a function-local macro that does `#[cfg(debug_assertions)] self.assert_validates()` and `if get_option!(dump_hir_iongraph) { passes.push(self.to_iongraph_pass($pass_name)); }`?